Fix stack walking for new EH

janvorli opened the pull request:
There are three issues caused by the new exception handling implementation causing GC holes. These were discovered by the CI GC stress tests and mostly occur only with GC stress 3 and tiered compilation disabled.

* There is an edge case when the StackFrameIterator was incorrectly setting the 
fShouldParentFrameUseUnwindTargetPCforGCReporting, which resulted in a 100% reproducible GC hole in the
baseservices\exceptions\unittests\ThrowInFinallyNestedInTry with GC stress C and tiered compilation off.
The fix is to delete code that is already present in an "if" branch above the change and that should really be executed only when the funclet parent is the caller of the actual handler frame.

* When GC occurs when a finally funclet is on the stack and then next GC happens when we are in the managed code of the second pass of the same exception handling, it introduces a GC hole. The latter GC still needs to report some slots in the parent of the finallys so that next GC scan when there is a finally again needs to find those alive. The current code works when the parent of the finally is the same as the parent of the actual exception handling catch funclet, but in other cases, the stack walk was not reporting the slots. It cannot just report all slots in the parent, as some of them would already be dead. Only slots that the funclets would report should be reported. 
The fix is to remember the PC, frame pointer and flags of the last GC reporting of a finally funclet and use that to report slots based off the frame pointer when no funclet of the same exception handling is found during a future GC.

* Last issue is related to the usage of SEH to pass through native frames when stack walking hits CallDescrWorkerInternal frame. The SEH stack unwinding doesn't update and doesn't even pass out the context pointers, so if a non-volatile register stored in a native frame contains a reference to an object that got relocated, a GC hole occurs. The Interop/Interop test was hitting that problem on Windows. 
The fix is to use context pointers from a skipped explicit frame to update context pointers in the current REGDISPLAY. When the SEH hits a managed frame in the 2nd pass, we restart exception handling on the context of the managed frame, so the explicit frame that was on the transition boundary between that managed frame and the just unwound native frames contains the context pointers we need.

janvorli committed:
Fix stack walking for new EH

There is an edge case when the StackFrameIterator was incorrectly
setting the fShouldParentFrameUseUnwindTargetPCforGCReporting, which
resulted in a 100% reproducible GC hole in the
baseservices\exceptions\unittests\ThrowInFinallyNestedInTry with GC
stress C and tiered compilation off.
The fix is to delete code that is already present in an "if" branch
above the change and that should really be executed only when the
funclet parent is the caller of the actual handler frame.

janvorli committed:
Testing only - enable the new EH for excercising it

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -1510,6 +1512,13 @@ void GcInfoDecoder::ReportRegisterToGC(  // AMD64
 {
     GCINFODECODER_CONTRACT;
 
+#ifndef NATIVEAOT
```
I would not bother with ifdefing tiny bits of code like this for native AOT.

jkotas replied:
If this is fixing build break, I would rather define a dummy `ReportFPBasedSlotsOnly` in native AOT. It can be defined to be `0` that should allow the compiler to eliminate this branch for native AOT.

janvorli replied:
This was just for my testing purposes, I didn't want to merge that in.

janvorli replied:
Actually, somehow CI has screwed this PR. For some reason, the description of the change I've added is missing and the commit doesn't contain the cleanup stuff. I need to check what happened and re-commit what I have locally.

janvorli replied:
Ah sorry, I was commenting on the log above. This is obviously needed. Yes, this was here to fix NativeAOT build break. I'll modify it the way you've suggested.

jkotas commented on src/coreclr/vm/gcinfodecoder.cpp:
```diff
@@ -680,7 +680,9 @@ bool GcInfoDecoder::EnumerateLiveSlots(
     {
-        LOG((LF_GCROOTS, LL_INFO100000, "Not reporting this frame because it was already reported via another funclet.\n"));
+#ifndef NATIVEAOT        
+        STRESS_LOG0(LF_GCROOTS, LL_INFO100, "Not reporting this frame because it was already reported via another funclet.\n");
```
I think that `LL_INFO100` too verbose log level for GC roots reporting.

janvorli replied:
Reverted this change as it was testing only

janvorli committed:
Cleanup, reflecting feedback and fixing one last issue

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
The fix for the last issue needs to be a bit involved, it seems - it is hitting an assert I've added there, so there seem to be a case when a skipped frame isn't one that needs register updates or it doesn't point to the managed frame. I need to check what is such case.

jkotas commented on src/coreclr/inc/gcinfodecoder.h:
```diff
@@ -186,6 +186,7 @@ enum ICodeManagerFlags
     NoReportUntracked
                     =   0x0080, // EnumGCRefs/EnumerateLiveSlots should *not* include
                                 // any untracked slots
+    ReportFPBasedSlotsOnly = 0, // Unused by nativeaot, the gc info decoder checks that. Set to 0 to let the compiler optimize out the related code.
```
```suggestion
    ReportFPBasedSlotsOnly
                    =   0x0200, // EnumGCRefs/EnumerateLiveSlots should only include
                                // slots that are based on the frame pointer
```
I assumed that the redundant definition is going to be somewhere under nativeaot. If it is here, I would make it match the other definition and live with the tiny bit of extra code.

janvorli committed:
Reflect PR feedback and fix asserts

janvorli committed:
Filter out InlinedCallFrame from updating the pointers

InlinedCallFrame needs to be excluded from updating the context
pointers, as it restores PC / SP to a different location than the
callsite of the pinvoke and it doesn't update the context pointers
anyways.

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr outerloop

janvorli commented:
Only two GC stress legs have failures due to timeout in a test
arm32: JIT/Directed/tailcall/more_tailcalls/more_tailcalls.cmd
arm64: Loader/binding/tracing/BinderTracingTest.ResolutionFlow/BinderTracingTest.ResolutionFlow.cmd

The arm32 happens even without the new EH enabled - see #99410
The arm64 seems to be an existing issue too - see #97735

AndyAyersMS commented:
https://github.com/dotnet/perf-autofiling-issues/issues/31934
